src/abstract-api-adapter.coffee

```diff
@@ -0,0 +1,103 @@
+request = require 'request'
+try
+	{ Adapter } = require 'hubot'
+catch
+	prequire = require 'parent-require'
+	{ Adapter } = prequire 'hubot'
+
+class AbstractAPIAdapter extends Adapter
+	constructor: ->
+		super
+		@interval = 1000
+		@lastReport = null
+		@allowedToPoll = true
+		throw new TypeError('Abstract class')
+
+	run: ->
+		setInterval @maybePoll, @interval
+		@emit 'connected'
+
+	maybePoll: =>
+		if @allowedToPoll
+			@allowedToPoll = false
+			@poll()
+
+	# Execute and callback a series of paged requests until we run out of pages or a filter rejects
+	getUntil: (options, each, filter, done) ->
+		request.get? options, (error, response, body) =>
+			if error or response?.statusCode isnt 200
+				if response?.statusCode in [429, 503]
+					@report { error: error, html: response?.statusCode, url: options.url }
+					# Retry if the error code indicates temporary outage
+					@getUntil options, filter, each, done
+				else
+					@robot.logger.error response?.statusCode + '-' + options.url
+					@processDone error, null, { error: error, html: response?.statusCode, url: options.url }, done
+			else
+				responseObject = JSON.parse(body)
+				results = @extractResults responseObject
+				for result in results
+					if (not filter?) or filter result
+						each result
+					else
+						@processDone null, results, { html: response?.statusCode, url: options.url }, done
+						return
+				if @extractNext responseObject
+					options.url = @extractNext responseObject
+					@getUntil options, filter, each, done
+				else
+					@processDone null, results, { html: response?.statusCode, url: options.url }, done
+
+	processDone: (error, response, report, done) =>
+		@allowedToPoll = true
+		@report report
+		done?(error, response)
+
+	extractResults: -> throw new TypeError('Abstract method')
+
+	extractNext: -> throw new TypeError('Abstract method')
+
+	report: (obj) =>
+		report = JSON.stringify(obj)
+		if @lastReport isnt report
+			@lastReport = report
+			@robot.logger.debug report
+
+	###*
+	* Given a set of ids make best effort to publish the text and send a new set of ids to the callback
+	* @param {string} text to publish
+	* @param {object} ids to use.  ids = {user, flow, thread?}
+	* @param {function} function to receive (error, new ids object).  ids = {thread, comment?}
+	###
+	postUsing: (text, ids, callback) =>
```

Comment:
Is there a reason we're not using promises here, and for `done` in `getUntil`/`processDone`? It'd make this code and code using this tidier and more reliable (done right, you should get automatic error -> promise error propagation, rather than having to ensure we're guaranteed to call the callback in every case)